docs/framework-design.md

```diff
@@ -0,0 +1,139 @@
+* To just build the library, if the platform-independent generated files are present, you only need the `include` directory and the directories containing library C files (`library`, `3rdparty`, `core`, `drivers` depending on the repository and desired features).
+
+#### Requirements for development in consuming repositories
+
+Consuming repositories must have the framework repository as a Git submodule for development work and CI scripts.
+
+Compared with pre-framework tooling, this means that Git submodules must be enabled. This requires an explicit step in many Git APIs (e.g. running `git submodule update --init` after `git init`, or passing `--recurse-submodules` to `git checkout`).
+
+#### Requirements for processes involving consuming repositories
+
+Release archives must include the content of the framework repository.
+
+#### Requirements for tooling in consuming repositories
+
+Consuming repositories may assume that the `framework` submodule is present wherever they assume a Git checkout.
+
+Consuming repositories may assume that the content of the `framework` directory is present anywhere where they would normally assume that all files are present. In particular, this allows the use of framework files for:
+
+* Generating configuration-independent files (e.g. `make generated_files`), including the ones in the `library` directory.
+* `make lib` (with GNU make or CMake) from a pristine checkout (because this involves `make generated_files`).
+* `make test` (even if all the tests have been built).
+
+Consuming repositories must not assume that the framework is present when merely building the library. In particular:
+
+* Our provided build scripts (e.g. `library/Makefile`, `library/CMakeLists.txt`) must not require any files from `framework` when compiling the library.
+* It's ok to have a file in `library` with a make dependency on a framework file, as long as the build works when the framework file is missing. This allows `make lib` to work as long as the generated files are present.
+* Library source files must not rely on headers from the framework.
+
+#### Requirements for users of consuming repositories
+
+Corresponding to the requirements on the repository above:
+
+* Contributors need the framework submodule.
+* Users who wish to run full CI tests need the framework submodule.
+* Users who want to build or run tests need the `framework` directory content.
+* Users who merely want to build the library, and who have the configuration-independent files already generated, do not need the `framework` directory content.
+
+## Contents of the framework repository
+
+### Criteria for inclusion
+
+In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and should be split into a shared part and a non-shared part.
```

Comment:
`tests/suites/test_suite_*` has a lot of code that's shared between branches, but also a lot of code that isn't shared, and tends to have differences inside individual functions and test cases. In particular, 2.28 code has legacy crypto preprocessor guards, 3.6 code has a lot of internal guards like `MBEDTLS_{MD,PK,CIPHER}_{HAVE,CAN}_xxx`, and 4.0 code will have `PSA_WANT`.

We do have places for the parts of `tests/suites/test_suite_*` that can be shared: test generator scripts when a lot of data follows a common model (which usually needs to be fined-tuned to the set of cryptographic mechanisms and API functions available in each branch), and `tests/src/*` for code that is usable in a generic way. This fulfills the “significant proportion of shared content and [MISSING PHRASE] should be split”.

Reply:
Do we gain anything (an increase in how much we can share) if we add a layer of abstraction to allow the guards to be in the separate branches but with the test cases themselves in the framework?

Reply:
Yes, we do gain some sharing possibilities, at the expense of some complexity. I expect that this complexity will be worth it for some auxiliary test functions. I don't think sharing the unit test cases between branches will be worth it, however. (_Some_ SSL test cases, maybe.) For example, I don't think it would make sense to have RSA test cases in the framework repository, but EdDSA test cases only in TF-PSA-Crypto because older maintained branches don't have EdDSA.